### DISCUSSION TITLE
Testing shouldn't be a second thought

### DISCUSSION THREAD

**Discussion by raphaelbadia** [vercel/next.js#44326] [App Router]
Hello,

I find that Nextjs projects are getting harder to test over the years. (Been using it since 2019)

The framework critically lacks official guidance about testing, a [simple search](https://github.com/vercel/next.js/discussions?discussions_q=testing) with the keyword `test` in the discussions tab of this repository shows people struggling with : 
- testing a component using `next/header`
- testing async components
- testing components using router, links
- testing components using `next/image`
- testing middlewares

When I learned testing, it was recommended to write a lot of integration tests (for example using react-testing-library to mock network requests and tests different happy paths, along with error and loading paths). Integration testing being supposedly a good tradeoff compared to slow, costly e2e tests. 

Now with next 13 and the app directory I feel like that the only way to test our apps will be e2e.

I wish that testing were more taken into account than a [simple page with basic examples](https://github.com/vercel/next.js/discussions/28173) ...

**Comment by GabenGar** [comment]
The only way to test web apps is e2e in the first place. `react-testing-library` and the likes which promise to replace proper e2e with just integration tests are the false shepherds of the real TDD, since they basically print checkmarks for their own abstraction of DOM. 
And yes, it does mean testing frontend code is pain in the ass, as it requires the most complicated testing approach. Said approach also not trivial to set up, because it can't be reduced to testing getters/setters line-by-line with a checkmark at the end.

**Reply by raphaelbadia** [reply]
I think it’s still a good trade off to test error and loading scenarios because e2e tests are super slow to run

**Reply by GabenGar** [reply]
But then you aren't really testing error and loading scenarios, since they run in some sort of framework's sandboxed impression of the runtime, not even close to the real one. Which in turn doesn't give you a working guarantee tests supposed to give you. At this point manual testing is more effective at doing its job.
You just have to decide if you write tests to catch errors early or for ritualistic purposes. 
In case of the former you don't need to write glorified getter/setter tests for every single scenario, but only for critical paths. And those do not require rerunning every second, but only at post-build/pre-deploy stages, aka a perfect window for e2e tests.
For the latter nextjs doesn't need to do anything, you just need to find a way to maximize green checkmarks per line of code and that sweet 100% coverage. Which might or might not be related to testing the underlying code.
